src/frontend/js/pages/TeacherDashboardContractsLayout/hooks/useDownloadContractArchive/index.tsx

```diff
@@ -0,0 +1,123 @@
+import { useEffect, useState } from 'react';
+import useContractArchive from 'hooks/useContractArchive';
+import { useContracts } from 'hooks/useContracts';
+import { ContractState } from 'types/Joanie';
+import { PER_PAGE } from 'settings';
+import useCheckContractArchiveExists from '../useCheckContractArchiveExists';
+import {
+  storeContractArchiveId,
+  unstoreContractArchiveId,
+  getStoredContractArchiveId,
+  isStoredContractArchiveIdExpired,
+} from './contractArchiveLocalStorage';
+
+export enum ContractDownloadStatus {
+  IDDLE = 'iddle',
+  PENDING = 'pending',
+  READY = 'ready',
+}
+
+interface UseTeacherContractsBulkDownloadProps {
+  organizationId: string;
+}
+
+const useDownloadContractArchive = ({ organizationId }: UseTeacherContractsBulkDownloadProps) => {
+  const {
+    items: contracts,
+    states: { isFetched },
+  } = useContracts({
+    organization_id: organizationId,
+    signature_state: ContractState.SIGNED,
+    page: 1,
+    page_size: PER_PAGE.teacherContractList,
+  });
+
+  const {
+    methods: { get: getArchive, create: createArchive },
+  } = useContractArchive();
+
+  const [contractArchiveId, setContractArchiveId] = useState<string | null>(null);
+  const { isContractArchiveExists, doCheckArchiveExists } = useCheckContractArchiveExists({
+    contractArchiveId,
+  });
+
+  const [isDownloadRequest, setIsDownloadRequest] = useState(false);
+  const [contractDownloadStatus, setContractDownloadStatus] = useState<ContractDownloadStatus>(
+    ContractDownloadStatus.IDDLE,
+  );
+
+  const onArchiveGenerationStart = (archiveId: string) => {
+    setContractArchiveId(archiveId);
+    storeContractArchiveId(archiveId);
+
+    doCheckArchiveExists(archiveId);
+  };
+
+  const onArchiveDownloadDone = () => {
+    setIsDownloadRequest(false);
+    setContractArchiveId(null);
+    unstoreContractArchiveId();
+  };
+
+  const doDownloadArchive = async () => {
+    if (isContractArchiveExists && contractArchiveId !== null) {
+      await getArchive(contractArchiveId);
+      onArchiveDownloadDone();
+      return;
+    }
+
+    // When archive is ready and download is requested
+    // a useEffect will trigger the download.
+    setIsDownloadRequest(true);
+
+    if (contractArchiveId === null) {
+      const archiveId = await createArchive(organizationId);
+      onArchiveGenerationStart(archiveId);
+    }
+  };
+
+  // Reset download status when contractArchiveId in local storage have been stored too long
+  useEffect(() => {
+    if (isStoredContractArchiveIdExpired()) {
+      onArchiveDownloadDone();
```

Comment:
Maybe the semantic here is a bit counter intuitive, instead `onArchiveDownloadDone` could be renamed as `clearArchive` on something like this, WDYT?